internal/controllers/machineset/machineset_controller.go

```diff
@@ -857,14 +864,28 @@ func (r *Reconciler) syncReplicas(ctx context.Context, s *scope) (ctrl.Result, e
 // There are small differences in how we calculate the Machine depending on if it
 // is a create or update. Example: for a new Machine we have to calculate a new name,
 // while for an existing Machine we have to use the name of the existing Machine.
-func (r *Reconciler) computeDesiredMachine(machineSet *clusterv1.MachineSet, existingMachine *clusterv1.Machine) *clusterv1.Machine {
+func (r *Reconciler) computeDesiredMachine(machineSet *clusterv1.MachineSet, existingMachine *clusterv1.Machine) (*clusterv1.Machine, error) {
+	nameTemplate := "{{ .machineSet.name }}-{{ .random }}"
+	if machineSet.Spec.MachineNamingStrategy != nil && machineSet.Spec.MachineNamingStrategy.Template != "" {
+		nameTemplate = machineSet.Spec.MachineNamingStrategy.Template
+		// This should never happen as this is validated on admission.
+		if !strings.Contains(nameTemplate, "{{ .random }}") {
+			return nil, errors.New("cannot generate MachineDeployment machine name: {{ .random }} is missing in machineNamingStrategy.template")
+		}
+	}
+
+	generatedMachineName, err := topologynames.MachineSetMachineNameGenerator(nameTemplate, machineSet.Spec.ClusterName, machineSet.Name).GenerateName()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate name for MachineSet machine")
```

Comment:
```suggestion
		return nil, errors.Wrap(err, "failed to generate Machine name")
```
Same in the KCP implementation